HLTriggerOffline/Muon/src/HLTMuonValidator.cc

```diff
@@ -63,86 +51,92 @@ class HLTMuonValidator : public DQMEDAnalyzer {
   HLTMuonPlotter::L1MuonMatcherAlgoForDQM const l1tMuonMatcherAlgo_;
 };
 
-//////////////////////////////////////////////////////////////////////////////
-//////// Namespaces, Typedefs, and Constants /////////////////////////////////
+HLTMuonValidator::HLTMuonValidator(const edm::ParameterSet &pset)
+    : pset_(pset),
+      hltProcessName_(pset.getParameter<std::string>("hltProcessName")),
+      hltPathsToCheck_(pset.getParameter<std::vector<std::string>>("hltPathsToCheck")),
+      triggerEventToken_(consumes(edm::InputTag("hltTriggerSummaryRAW"))),
+      genParticlesToken_(consumes(pset.getParameter<std::string>("genParticleLabel"))),
+      recoMuonsToken_(consumes(pset.getParameter<std::string>("recMuonLabel"))),
+      l1tMuonMatcherAlgo_(pset, consumesCollector()) {}
 
-using namespace std;
-using namespace edm;
-using namespace reco;
-using namespace trigger;
+void HLTMuonValidator::fillLabels(std::string const &path,
+                                  std::vector<std::string> &moduleLabels,
+                                  std::vector<std::string> &stepLabels) const {
+  auto const &hltFilters = hltConfig_.saveTagsModules(path);
 
-typedef vector<string> vstring;
+  moduleLabels.clear();
+  moduleLabels.reserve(hltFilters.size());
 
-//////////////////////////////////////////////////////////////////////////////
-//////// Class Methods ///////////////////////////////////////////////////////
+  stepLabels.clear();
+  stepLabels.reserve(hltFilters.size() + 1);
 
-HLTMuonValidator::HLTMuonValidator(const ParameterSet &pset)
-    : pset_(pset),
-      hltProcessName_(pset.getParameter<string>("hltProcessName")),
-      hltPathsToCheck_(pset.getParameter<vstring>("hltPathsToCheck")),
-      triggerEventToken_(consumes(edm::InputTag("hltTriggerSummaryRAW"))),
-      genParticlesToken_(consumes(pset.getParameter<string>("genParticleLabel"))),
-      recoMuonsToken_(consumes(pset.getParameter<string>("recMuonLabel"))),
-      l1tMuonMatcherAlgo_(pset, consumesCollector()) {}
+  for (auto const &module : hltFilters) {
+    if (module.find("Filtered") == std::string::npos)
+      continue;
 
-vector<string> HLTMuonValidator::moduleLabels(string path) {
-  vector<string> modules = hltConfig_.moduleLabels(path);
-  vector<string>::iterator iter = modules.begin();
+    auto const step_label = stepLabel(module);
+    if (step_label.empty() or std::find(stepLabels.begin(), stepLabels.end(), step_label) != stepLabels.end())
+      continue;
 
-  while (iter != modules.end())
-    if (iter->find("Filtered") == string::npos)
-      iter = modules.erase(iter);
-    else
-      ++iter;
+    moduleLabels.emplace_back(module);
+    stepLabels.emplace_back(step_label);
+  }
 
-  return modules;
-}
+  if (stepLabels.empty()) {
+    return;
+  }
+
+  if (stepLabels[0] != "L1" and std::find(stepLabels.begin(), stepLabels.end(), "L1") != stepLabels.end()) {
```

Comment:
This forbids cases where `"L1"` is present in `stepLabels`, but is not the first module found by `HLTMuonValidator` in the HLT Path (the case where `"L1"` is not present in `stepLabels` remains supported).